keps/sig-instrumentation/3077-contextual-logging/README.md

```diff
@@ -0,0 +1,974 @@
+when trying the same test invocation locally.
+
+When everything works, he wants `go test` to hide the log output to avoid
+blowing up the size of the CI log files.
+
+For each inner test case he adds a `NewTestContext(t)` invocation and uses the
+returned context and logger for that test case.
+
+#### Story 4
+
+Client developer Joan wants to use client-go in his application, but is [less
+interested in log messages from
+it](https://kubernetes.slack.com/archives/CG3518SFJ/p1634217685020400). Joan
+makes the log messages from client-go less verbose by creating a logger with
+`logger.V(1)` and passing that to the client-go code. Now a `logger.V(3).Info`
+call in client-go is the same as a `logger.V(4).Info` in the application and
+will not be shown for `-v=3`.
+
+
+### Risks and Mitigations
+
+#### Design
+
+##### Uninitialized logger
+
+One risk is that code uses an uninitialized `logr.Logger`, which would lead to
+nil pointer panics. This gets mitigated with the logr convention that a
+`logr.Logger` passed by value always must be usable. When the logger is optional,
+this needs to be indicated by passing a pointer.
+
+##### Logging during initialization
+
+Another problem is that log messages during program startup need to be handled
+somehow. We then can either:
+- discard the message, which would break the current `klog.Error(...);
+  os.Exit(1)` code pattern when an error occurs during initialization log
+- log with some default configuration (for example, as text)
+
+In both cases we have the problem that a logger retrieved by some init code
+before logging initialization will continue to use that behavior (discard or
+log with default configuration) even after logging gets initialized.
+
+Ideally, no log messages should be emitted before the program is done with
+setting up logging as intended. This is already [problematic
+now](https://github.com/kubernetes/kubernetes/issues/100152) because output may
+change from text (the current klog default) to JSON (once initialized). There
+will be no automatic mitigation for this. Such log calls will have to be found
+and fixed manually, for example by [passing the error back to `main`
+instead](https://github.com/kubernetes/kubernetes/pull/104774), which is part
+of an effort to [remove the dependency on logging before an unexpected program
+exit](https://github.com/kubernetes/kubernetes/issues/102231).
+
+To support identifying log messages that need to be fixed, the default logger
+during program startup will be using the klog text output instead of discarding
+the output and a `INIT-LOG-FIXME` prefix will be inserted into the log output.
+
+##### Performance overhead
+
+Retrieving a logger from a context on each log call will have a higher overhead
+than using a global logger instance. The overhead will be measured for
+different scenarios. If a function uses a logger repeatedly, it should retrieve
+the logger once and then use that instance.
+
+#### Transition
```

Comment:
what is the impact on consumers of k8s.io/... libraries that currently set up klog/v2 in particular ways and expect k8s.io components to use it?

Reply:
Initially, and potentially for quite a while, k8s.io libraries will still depend and use klog even though the code gets rewritten to support contextual logging. During that phase, such third-party components will continue to work unchanged, i.e. they configure klog and the libs pick that up. Third-party components can already prepare for the removal of the klog dependency by replacing direct usage of klog with usage of k8s.io/component-base/logs and k8s.io/klogr.

At some point when we feel that downsteam is ready, we can modify k8s.io/klogr so that it no longer uses klog. Only then will components have to pass in a logger through the context. The prototype branch has a commit which implements that, just to verify that it'll work as envisioned.

Reply:
cluster-autoscaler doesn't need to migrate. There is a new `WithLogger` option for the framework, but that doesn't have to be used because the global klog will still work initially.

Reply:
How can downstream components like Cluster Autoscaler learn that they need to migrate? And how long will this initial phase take (i.e. how can we know that the downstream components are ready enough)?

Reply:
> How can downstream components like Cluster Autoscaler learn that they need to migrate?

I'll write a blog post and have a talk proposal pending for KubeCon.

> And how long will this initial phase take (i.e. how can we know that the downstream components are ready enough)?

We can search for "klog" usage in downstream components. Those that have migrated will not call it directly anymore. I'm open for suggestions for how broad that search should be at which graduation phase.

Then what should we (= SIG Instrumentation) do if downstream components are not getting migrated? Filing bugs may be possible, but preparing and testing PRs is probably going to be too much work.

Reply:
> We can search for "klog" usage in downstream components. Those that have migrated will not call it directly anymore. I'm open for suggestions for how broad that search should be at which graduation phase.

This works for OSS, but any closed source operators / tools won't be covered. Perhaps this is fine, but I think klog should start logging a warning at startup at some point in the deprecation phase.

> Then what should we (= SIG Instrumentation) do if downstream components are not getting migrated? Filing bugs may be possible, but preparing and testing PRs is probably going to be too much work.

Maybe a concrete deprecation plan would help? I.e. define phases of deprecation with defined minimal soak time (and/or exit criteria) for each and hang them somewhere everyone can see. Filing bugs could be (a part of) one of the phases, and the bug content could link to the deprecation plan.

Reply:
> Perhaps this is fine, but I think klog should start logging a warning at startup at some point in the deprecation phase.

Deprecating klog is not currently a goal of this KEP. The goal is just to remove the hard dependency on it and then to stop using it altogether in Kubernetes code.

But that raises a valid question: what is the fate of klog when we are done with that? Some users might decide to continue using it, and they can do that even with newer Kubernetes code (initialize klog, install klog/klogr as contextual logger).

IMHO we shouldn't cause unnecessary pain for consumers of our code and therefore keep klog as it is. How much support users of it then can still get is up for debate, though.

> Maybe a concrete deprecation plan would help? I.e. define phases of deprecation with defined minimal soak time (and/or exit criteria) for each and hang them somewhere everyone can see. 

The graduation criteria come pretty close already. I've extended it further to cover the things we discussed here (checking and notifying users of the code). The enhancement tracking issue description can be used to track the current status.